pepper/apps/trade_execution.c

```diff
@@ -43,37 +41,39 @@ void compute(struct In *input, struct Out *output) {
         struct Order order = orders[index];
         struct Volume volume = volumes[index];
         
-        if (fieldToInt(volume.sellVolume) > 0) {
-            // Verify volume has same ratio as prices
-            assert_zero((volume.buyVolume * prices[fieldToInt(order.buyToken)]) - (volume.sellVolume * prices[fieldToInt(order.sellToken)]));
+        // Verify volume has roughly same ratio as prices
+        field254 lhs = volume.buyVolume * prices[fieldToInt(order.buyToken)];
+        field254 rhs = volume.sellVolume * prices[fieldToInt(order.sellToken)];
+        field254 delta = lhs - rhs;
+        // Make sure |delta| < epsilon
+        assert_zero(isNegative((EPSILON*EPSILON) - delta) || isNegative((EPSILON*EPSILON) + delta));
```

Comment:
What is the reason for having EPSILON square?

Reply:
Epsilon is designed for rounding errors of a single decimal number (e.g. when summing each buyVolume and comparing to the total Volume). 
However, here we are multiplying two decimals (price * volume), thus squaring the rounding error we get from a single digit. We could alternatively set a high enough epsilon in the first place, but when there is no multiplication involved, we should not need such a hight value.